### PR TITLE
Implement address family affinity and prefer IPv4 by default

### PR DESCRIPTION
**Pull request by jchadwick-buf**
Implements address family policy. When the policy is set to `PreferIPv4` or `PreferIPv6`, any DNS result that has both IPv4 and IPv6 records will be filtered to only the preferred records; in cases where only IPv4 or IPv6 addresses are present, this has no impact. The default is set to `PreferIPv4` to help lower the likelihood of problems due to the fact that `httplb` lacks some kind of implementation of a "happy eyeballs" algorithm to perform IPv6 fallback.

It winds up being a little tricky to test this, but it's probably worthwhile; right now we can only test the DNS resolver in a fairly limited way (by relying on the fact that resolving an IP address is a no-op.) With this approach, we can test the DNS resolver more-or-less end-to-end.

### CODE REVIEW

File: resolver/resolver_test.go
Comment (jchadwick-buf) [reply resolver/resolver_test.go:91]: The fake `net.Resolver` part is actually pretty straightforward imo, and at 72 lines including error handling and other boilerplate it's not really that long. Would be nicer if you could flat-out fake results, but the two ways to do that are internal to Go standard library only.

You're right, though, this could just be done by making the affinity handling its own function. Probably should do that.

Do we really want to delete these tests entirely, though? I actually quite like having a proper end-to-end test of the DNS resolver that is somewhat less trivial. Since it really does everything as if we're in the real world, it shouldn't be very fragile either.

File: resolver/resolver.go
Comment (jchadwick-buf) [reply resolver/resolver.go:196]: As far as whether an IPv4 single-stack system would handle it, I think as long as the OS and network stack have IPv6 support in them to be able to handle the address the answer is broadly yes. As for whether or not that means this conditional belongs here... Probably not. I think I added it out of an abundance of caution, but there is no evidence that I'm aware of that people are sneaking IPv4 addresses into AAAA records or anything bizarre like that.

Comment (jchadwick-buf) [reply resolver/resolver.go]: Because we do unwrap it using `Unmap` I'm kinda unsure if we want to remove this check. If we remove this check it feels like we should remove the `Unmap` too.

File: resolver/resolver.go
Comment (jhump) [inline resolver/resolver.go:139]: I just had a thought: the affinity is only useful when combined with "ip" network, right?

Maybe we should _remove_ the `network` argument and expand the affinity settings to include two new ones: `RequireIPv4` and `RequireIPv6`. Under the hood, the latter two would result in "ip4" or "ip6" when actually making the `LookupNetIP` call.

File: resolver/resolver.go
Comment (jhump) [inline resolver/resolver.go:36]: Thinking about this more, I think it's a little odd for this choice to be the zero value. Maybe instead we order them (incorporating idea from other comment):
* `PreferIPv4`
* `RequireIPv4`
* `PreferIPv6`
* `RequireIPv6`
* `UseBothIPv4AndIPv6`

I changed `AllFamilies` to `UseBothIPv4AndIPv6` in the example list above since I think that would make call sites a littler more readable as to what that argument means. In that same vein, maybe rename the type to `AddressFamilyPolicy`?

WDYT?

File: resolver/resolver.go
Comment (jhump) [reply resolver/resolver.go]: Okay. I think it might be the most intuitive behavior (and easiest to document) to just skip this condition and skip the unmap below. That way we can simply say that "PreferIPv4" will prefer results from DNS A records and "PreferIPv6" will prefer results from DNS AAAA records, and we don't need to try to document a bunch nuance about 4-in-6 addresses.

File: resolver/resolver_test.go
Comment (jhump) [reply resolver/resolver_test.go:91]: Sure, leave 'em. I guess it was mostly an observation (sorry if unclear, but I tried to indicate in review summary that this was not a blocking comment).

Comment (jhump) [reply resolver/resolver_test.go:91]: If we ever get to doing a "happy eyeballs" implementation, we likely will want to re-use this, so we can re-use the `net.Dialer`'s implementation, and just have a mocked resolver that provides exactly the addresses we want...

File: resolver/resolver.go
Comment (jchadwick-buf) [reply resolver/resolver.go:36]: Hmmm, sure, that all sounds good. It seems like this makes things a little clearer/simpler from the user perspective in exchange for slightly overloading what we actually do on our end.

File: resolver/resolver.go
Comment (jchadwick-buf) [reply resolver/resolver.go]: Surprise! We actually can't. It breaks tests, because in fact there are cases where you WILL get IPv4-embedded IPv6 addresses, and we actually hit one in our tests (thank goodness.) So not only do I need to keep this logic, I also need to test it most likely.

Why? Well, because the trick we were _previously_ relying on was the short circuit that causes resolution of an IP address to be more or less a no-op, not actually hitting either the Go or system resolver. But why? [Well](https://cs.opensource.google/go/go/+/refs/tags/go1.22.2:src/net/lookup.go;l=309-311)....

```go
	if ip, err := netip.ParseAddr(host); err == nil {
		return []IPAddr{{IP: IP(ip.AsSlice()).To16(), Zone: ip.Zone()}}, nil
	}
```

But why? I'm not sure. I dug unnecessarily deep and I still don't really have the answer.

Unfortunately, since this is a super common case, I think we absolutely must handle it. So... nevermind.

File: resolver/resolver.go
Comment (jchadwick-buf) [reply resolver/resolver.go]: Done, now `policy` replaces `network` and `affinity`.

File: resolver/resolver.go
Comment (jchadwick-buf) [reply resolver/resolver.go]: Done.

File: resolver/resolver_test.go
Comment (jchadwick-buf) [reply resolver/resolver_test.go:91]: Now that we have made policy handle the `network` setting as well, I think testing end-to-end makes even more sense, so I went ahead and did that.

File: resolver/resolver.go
Comment (jhump) [reply resolver/resolver.go]: Okay. So I think the only thing squirrelly that might happen is if a DNS result includes an A record _and_ an AAAA record that is a 4-in-6-style address for the same thing as the A record. We'd end up establishing double connections to it, which isn't necessarily bad and is likely pretty defensible behavior.

File: client.go
Comment (jhump) [inline client.go:40]: I think we want this default to be controllable via env var. Maybe something like `HTTPLB_DEFAULT_ADDRESS_FAMILY_POLICY`. We can document the exact values the env var needs to have either in `NewClient` or maybe it would be better in `doc.go` where it talks about default behavior.

**Review by jhump — COMMENTED** [review]
LGTM, but we still need some updates to other doc comments and also it would be nice to allow the default to be controllable via env var (and use `PreferIPv4` if there is no such env var or if it doesn't have a valid value).

File: client.go
Comment (jchadwick-buf) [reply client.go:40]: Hmmm... To be honest, I'm not sold on this. There are exceptions where it may be useful (a la `GODEBUG`) but in the general case I see this as a code smell. It essentially adds an implicit "impure" configuration knobs that skip the application's own configuration mechanisms. If an application wants to create a "default" resolver with one setting changed, they have to replicate this behavior too.

If applications want this behavior to be controlled by environment variable, I think they ought to do it themselves.

If you insist I can add it, but is there really a good reason why reading and parsing environment variables is something `httplb` itself should do?

Comment (jhump) [reply client.go:40]: > skip the application's own configuration mechanisms

It would only be for the default, where the application has not configured anything (which will typically be unintentional, either an inadvertent omission or lack of foresight). If the application _does_ configure it, the env var would not override.

> If applications want this behavior to be controlled by environment variable, I think they ought to do it themselves.

I guess my worry is that this would never happen in practice. If they are adding code of this sort at all, then they are thinking about this and likely configuring things intentionally. More likely is that this isn't thought about or configured at all because the defaults "just worked" on day one. And then on day two, when some vendor pushes a DNS change, things break (such as in an IPv6 single-stack setup, where DNS is updated to suddenly start returning an A record). I guess I'm just used to providing knobs for certain scenarios that don't changing code, re-building, and re-deploying, as a dev ops instinct. 🤷 

I guess we can leave it out. But this is the only behavior where httplb will observably differ from net/http and could do so in a way that prevents successful client->server communication, so it seemed useful. To be honest, I don't place a lot of value on "purity", especially if it potentially compromises usability or developer experience.

Anyhow, if you really don't like it, leave it out. But we definitely _do_ need to update the docs about the change in default behavior.

**Review by jhump — APPROVED** [review]
Please update the docs about the default behavior before merging.

**Comment by jchadwick-buf** [comment]
> Please update the docs about the default behavior before merging.

I already did do this, and I can't find anywhere else where it's wrong. Am I missing something?

Regardless, I did add more documentation.

**Comment by jhump** [comment]
> I already did do this, and I can't find anywhere else where it's wrong. Am I missing something?

Nope. I left that comment just before you pushed the commit with doc updates. Updates look great! Thanks!

File: client.go
Comment (jchadwick-buf) [reply client.go:40]: I think I'll hold off on adding this. I can see the operational value of a feature like this, but in practice I think that it is least surprising to users if a library has one and only one way of configuring things, through the API, rather than environment variables. If we do want to make this sort of thing configurable through environment variables, it'd probably be worth considering _other_ things that might be worth exposing as environment variables and/or flags, possibly as a side-effect of importing an optional package, kind of like how some C++ libraries will have packages you can link to to get absl flags. I can see some arguments against this, too (it wouldn't help for people who didn't explicitly think of this, which is arguably the people who have the most to gain from the built-in ability to change a configuration knob at runtime) but ultimately I think that the better way to resolve this potential need is to make our IPv6 fallback work better in the future, so that users don't really have to think about it much more than they would for `net.Dial` and `http.DefaultClient`.

**Pull request merged**
